service/history/workflow/mutable_state_impl.go

```diff
@@ -491,10 +492,25 @@ func NewMutableStateInChain(
 	// carry over necessary fields from current mutable state
 	newMutableState.executionInfo.WorkflowExecutionTimerTaskStatus = currentMutableState.GetExecutionInfo().WorkflowExecutionTimerTaskStatus
 
+	// Copy completion callbacks to new run.
+	oldCallbacks := callbacks.MachineCollection(currentMutableState.HSM())
+	newCallbacks := callbacks.MachineCollection(newMutableState.HSM())
+	for _, node := range oldCallbacks.List() {
+		cb, err := oldCallbacks.Data(node.Key.ID)
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := cb.PublicInfo.GetTrigger().GetVariant().(*workflowpb.CallbackInfo_Trigger_WorkflowClosed); ok {
+			if _, err := newCallbacks.Add(node.Key.ID, cb); err != nil {
```

Comment:
Only thing I'm worried about is if this may error because we add the same callback twice in some path. Not sure if that can happen or not.

Reply:
AFAICT it can't happen. The only other places we add callbacks is in the ContinueAsNew path, which doesn't use this, and in `ApplyWorkflowExecutionStartedEvent` but for retries we don't put the callbacks on the started event, so I don't think they will get double added there either.